### ISSUE TITLE
DirectoryServicesCOMException - Unexpected error occurred while getting domain names information

### ISSUE THREAD

**Issue by Yvand** [Yvand/LDAPCP#87]
Randomly, LDAPCP fails to get domain names information and records the following error:

```
07/18/2019 13:58:53.78	w3wp.exe (0x22A0)	0x1AE8	LDAPCP	Configuration	1337	Unexpected	[LDAPCP] Unexpected error occurred while getting domain names information for LDAP connection LDAP://contoso.local/DC=contoso,DC=local: System.DirectoryServices.DirectoryServicesCOMException: An operations error occurred.  , Callstack:   
at System.DirectoryServices.DirectoryEntry.Bind(Boolean throwIfFail)    
at System.DirectoryServices.DirectoryEntry.Bind()    
at System.DirectoryServices.DirectoryEntry.get_AdsObject()    
at System.DirectoryServices.PropertyCollection.Contains(String propertyName)    
at ldapcp.OperationContext.GetDomainInformation(DirectoryEntry directory, String& domainName, String& domainFQDN)    
at ldapcp.LDAPCP.SetLDAPConnection(Uri currentContext, LDAPConnection ldapConnection)
```

As a consequence, entities that use dynamic domain name tokens are missing the domain name.
For example, entity value will be "\groupName" instead of "contoso.local\groupName"

**Comment by Yvand** [comment]
Good news, I found the root cause and upcoming version 14 will fix it. If you are interested, here are the details about the bug:

## Context
The issue occurs when LDAPCP uses the LDAP (AD) connection of the SharePoint server.
At some point LDAPCP gets the domain name by reading the LDAP property “distinguishedName” from the object DirectoryEntry:

https://github.com/Yvand/LDAPCP/blob/a5c164680f44fbf778446bc91b0f59f887f557a5/LDAPCP/LDAPCPConfig.cs#L1290

Reading this property triggers an authenticated LDAP call to the AD server, using the creds of the application pool account (LDAPCP does privilege elevation).

## Root cause
But this privilege elevation was missing in 2 entry points: in [LDAPCP.GetClaimTypeForUserKey()](https://github.com/Yvand/LDAPCP/blob/a5c164680f44fbf778446bc91b0f59f887f557a5/LDAPCP/LDAPCP.cs#L1916) and [LDAPCP.GetUserKeyForEntity()](https://github.com/Yvand/LDAPCP/blob/a5c164680f44fbf778446bc91b0f59f887f557a5/LDAPCP/LDAPCP.cs#L1948).
So from those methods, connection to AD server was attempted as “NT AUTHORITY\IUSR” (the default anonymous account), which of course failed: lsass was issuing a Kerberos TGS to SPN “krbtgt/NT AUTHORITY” (assuming NT AUTHORITY was an AD domain name).
This is what caused the DirectoryServicesCOMException.

## Solution
I added the missing privilege elevation in method LDAPCP.GetUserKeyForEntity():

https://github.com/Yvand/LDAPCP/blob/a5c164680f44fbf778446bc91b0f59f887f557a5/LDAPCP/LDAPCP.cs#L1953-L1956

But doing so in LDAPCP.GetClaimTypeForUserKey() is not possible: it causes a StackOverflowException. 
So to fully fix the issue I implemented a different way to read the domain name in this scenario, and little other improvements.
The commits that fix the bug are those with #87 in their title:

- https://github.com/Yvand/LDAPCP/commit/653e1d9f17f301474b41df089fd23b2569d283cb
- https://github.com/Yvand/LDAPCP/commit/8b024e268a4b499c97916e39674dde1430b0ffdb
- https://github.com/Yvand/LDAPCP/commit/c7daef4d197b844add2cdc0a75e1c916ad724b4f
- https://github.com/Yvand/LDAPCP/commit/a5c164680f44fbf778446bc91b0f59f887f557a5

Upcoming version 14 will contain the fix. I will publish it as soon as possible, with hopefully a pre-release today if unit tests pass.

**Comment by Yvand** [comment]
The pre-release that fixes this bug is https://github.com/Yvand/LDAPCP/releases/tag/14.0.20190821.952
